### PR TITLE
VIP FS: Add append mode support to streamwrapper

### PR DESCRIPTION
**Pull request by hanifn**
## Description

Add support for append mode file operations into the streamwrapper class.

Currently using `fopen()` with `a` or `a+` mode or `file_put_contents()` with the `FILE_APPEND` flag does not work as expected. This should fix that.

## Checklist

Please make sure the items below have been covered before requesting a review:

- [x] This change works and has been tested locally (or has an appropriate fallback).
- [x] This change works and has been tested on a Go sandbox.

## Steps to Test

Outline the steps to test and verify the PR here.

Example:

1. Sandbox a test site
1. `wp shell`
1. Create new file: `file_put_contents( 'vip://wp-content/uploads/append-test.txt', 'new file' );`
1. Append to file: `file_put_contents( 'vip://wp-content/uploads/append-test.txt', ' appended', FILE_APPEND );`
1. Verify file contents is now `new file appended`.

### CODE REVIEW

**Comment by nickdaugherty** [comment]
Crap, somehow missed the Review Request notification for this...looking now...

File: files/class-vip-filesystem-stream-wrapper.php [hunk 1]
```diff
@@ -650,4 +680,69 @@ protected function close_handler() {
...
+	*
+	* @since   1.0.0
+	* @access  private
+	* @param   string    $path   Path to file
+	* @param   string    $mode   fopen mode
+	*
+	* @return  bool
+	 */
+	private function validate( $path, $mode ) {
+		if ( ! in_array( $mode, [ 'r', 'w', 'a', 'x' ], true ) ) {
```
Comment (nickdaugherty) [inline files/class-vip-filesystem-stream-wrapper.php]: Will `$mode` _always_ have a value?

File: files/class-vip-filesystem-stream-wrapper.php [hunk 2]
```diff
@@ -122,6 +131,11 @@ public function register() {
 	 */
 	public function stream_open( $path, $mode, $options, &$opened_path ) {
 		$path = $this->trim_path( $path );
+		$mode = rtrim( $mode, 'bt' );
```
Comment (nickdaugherty) [inline files/class-vip-filesystem-stream-wrapper.php]: What does the `rtrim()` for `bt` do?

File: files/class-vip-filesystem-stream-wrapper.php [hunk 3]
```diff
@@ -220,6 +235,11 @@ public function stream_flush() {
 			return false;
 		}
 
+		if ( 'r' === $this->mode ) {
+			// No writes in 'read' mode
+			return false;
```
Comment (nickdaugherty) [inline files/class-vip-filesystem-stream-wrapper.php]: Should this `trigger_error()` too?

File: files/class-vip-filesystem-stream-wrapper.php (hunk 1 above)
Comment (nickdaugherty) [reply files/class-vip-filesystem-stream-wrapper.php]: PHP supports more modes than this:

https://www.php.net/manual/en/function.fopen.php

Specifically `r+`, `w+`, `a+`, `x+`, `c`, `c+`, and `e` (not really used). We don't necessarily need to implement those all right away, but we should at least document what is and isn't supported (and probably move the supported modes to a constant).

Comment (hanifn) [reply files/class-vip-filesystem-stream-wrapper.php]: > Will `$mode` _always_ have a value?

Yes as `$mode` is from `fopen()` and is a required argument.

Comment (hanifn) [reply files/class-vip-filesystem-stream-wrapper.php]: > We don't necessarily need to implement those all right away, but we should at least document what is and isn't supported (and probably move the supported modes to a constant).

This is a good idea. I'll add that in. Also, we're ignoring the `+` modes as in our current implementation, all the file handles are read/write capable.

File: files/class-vip-filesystem-stream-wrapper.php (hunk 2 above)
Comment (hanifn) [reply files/class-vip-filesystem-stream-wrapper.php]: `b` is force binary mode and `t` is Windows-only text mode. It's used along with the regular modes of `fopen()` like so: `rb` or `r+b`. Both of which does not apply, I believe, to the functionality of the stream wrapper thus we're stripping them here.

File: files/class-vip-filesystem-stream-wrapper.php (hunk 3 above)
Comment (hanifn) [reply files/class-vip-filesystem-stream-wrapper.php]: I originally did not think so but a `trigger_error()` could help with debugging so I'll add that in.

File: files/class-vip-filesystem-stream-wrapper.php (hunk 1 above)
Comment (hanifn) [reply files/class-vip-filesystem-stream-wrapper.php]: Changes made in a145ce99ed9c3643e9f5db25e440a0894d283007

File: files/class-vip-filesystem-stream-wrapper.php (hunk 3 above)
Comment (hanifn) [reply files/class-vip-filesystem-stream-wrapper.php]: Changes made in f60d94b3aba76cc7357adb618ed9e13448ab9063
I've also added the same check to `stream_write()`

**Comment by hanifn** [comment]
Hmmm... `A8C_Files_ImageSizes_Test` tests are failing

File: files/class-vip-filesystem-stream-wrapper.php (hunk 1 above)
Comment (nickdaugherty) [reply files/class-vip-filesystem-stream-wrapper.php]: Awesome!

**Comment by nickdaugherty** [comment]
> Hmmm... A8C_Files_ImageSizes_Test tests are failing

Once we figure that out, how difficult would it be to add a test for the new `a` mode support?

**Comment by mjangda** [comment]
The test failures are because of an upstream change in 5.3 (core adds two new image sizes)

**Comment by chrean** [comment]
@hanifn what's the status on this one? Do we need to add tests as @nickdaugherty suggested, or?

**Comment by hanifn** [comment]
@chrean TBH, I'm not exactly sure how to test this in PHPUnit. Should we just merge this and then figure that out later?

File: files/class-vip-filesystem-stream-wrapper.php [hunk 4]
```diff
@@ -650,4 +706,69 @@ protected function close_handler() {
...
+	 * Validates the provided stream arguments for fopen.
+	*
+	* @since   1.0.0
+	* @access  private
+	* @param   string    $path   Path to file
+	* @param   string    $mode   fopen mode
+	*
+	* @return  bool
+	 */
+	private function validate( $path, $mode ) {
```
Comment (nickdaugherty) [inline files/class-vip-filesystem-stream-wrapper.php]: If we make this a `public` method, it becomes testable with PHPUnit.
```suggestion
	public function validate( $path, $mode ) {
```

**Comment by hanifn** [comment]
r681-stacks

**Pull request merged**